### PR TITLE
Add ability for bot to get xkcd comics

### PR DESCRIPTION
**Pull request by namemcguffin**
Adds ability for bot to fetch xkcd comics

## Description
Added one new function, ?xkcd, which sends an embed containing the xkcd comic requested, or the latest one if no number was specified

## Motivation and Context
Porting functionality from Mose

## How Has This Been Tested?
Tested in the idoneam test server

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: cogs/memes.py [hunk 1]
```diff
@@ -169,6 +173,43 @@ def pyramidy(n, m):
...
+            xkcd_req = get("https://xkcd.com/")
+        else:
+            try:
+                xkcd_req = get(f"https://xkcd.com/{int(xkcd_command)}")
+            except ValueError:
+                await ctx.send(
+                    f"invalid input: {xkcd_command} does not parse to an integer"
+                )
+                return
+        if xkcd_req.status_code == 200:
```
Comment (davidlougheed) [inline cogs/memes.py:196]: early return if not 200 since its the exception case

File: cogs/memes.py [hunk 2]
```diff
@@ -19,9 +19,13 @@
 
 # discord-py requirements
 from discord.ext import commands
+from discord import Embed
 
 # Other utilities
 import random
+from requests import get
```
Comment (davidlougheed) [inline cogs/memes.py:26]: i would probably just do `import requests` and use `requests.get` here to make the global namespace a bit less wierd

File: cogs/memes.py [hunk 3]
```diff
@@ -19,9 +19,13 @@
...
+from bs4 import BeautifulSoup
+from re import findall
```
Comment (davidlougheed) [inline cogs/memes.py:28]: same here for import

File: cogs/memes.py [hunk 4]
```diff
@@ -19,9 +19,13 @@
 
 # discord-py requirements
 from discord.ext import commands
+from discord import Embed
```
Comment (davidlougheed) [inline cogs/memes.py:22]: same here for import

File: cogs/memes.py [hunk 5]
```diff
@@ -169,6 +173,43 @@ def pyramidy(n, m):
...
+                )
+                return
+        if xkcd_req.status_code == 200:
+            xkcd_soup = BeautifulSoup(xkcd_req.content, "html.parser")
+            xkcd_img_soup = xkcd_soup.find("div", attrs={
+                "id": "comic"
+            }).find("img")
+            xkcd_embed = Embed(
+                title=
+                f"{xkcd_img_soup['alt']} (number: {findall(r'^https://xkcd.com/([1-9][0-9]*)/$', xkcd_soup.find('meta', property='og:url')['content'])[0]})"
```
Comment (davidlougheed) [inline cogs/memes.py:203]: this line is too long

**Review by davidlougheed — CHANGES_REQUESTED** [review]
probably don't need to preface all the variables with `xkcd_`, since theyre scoped in the function anyway

File: cogs/memes.py [hunk 6]
```diff
@@ -169,6 +173,44 @@ def pyramidy(n, m):
...
+                xkcd_req = requests.get(
+                    f"https://xkcd.com/{int(xkcd_command)}")
+            except ValueError:
+                await ctx.send(
+                    f"invalid input: {xkcd_command} does not parse to an integer"
+                )
+                return
+        if xkcd_req.status_code != 200:
+            await ctx.send(
+                f"xkcd number {xkcd_command} could not be found (request returned {xkcd_req.status_code})"
```
Comment (le-potate) [inline cogs/memes.py:199]: I'd write (#4) instead of (number: 4) here too

File: cogs/memes.py [hunk 7]
```diff
@@ -169,6 +173,44 @@ def pyramidy(n, m):
...
+        Enjoy a nice xkcd comic with some strangers on the internet!
+        If no issue number is passed, returns a random xkcd.
+        Valid options are: nothing, 'latest' and a positive integer greater than one and at most equal to the latest issue number.
+        """
+        await ctx.trigger_typing()
+        if xkcd_command is None:
+            xkcd_req = requests.get("https://c.xkcd.com/comic/random")
+        elif xkcd_command == "latest":
+            xkcd_req = requests.get("https://xkcd.com/")
+        else:
```
Comment (le-potate) [inline cogs/memes.py:188]: check if the number is greater than 1

File: cogs/memes.py [hunk 8]
```diff
@@ -169,6 +173,44 @@ def pyramidy(n, m):
 
         await ctx.send("**\n{}**".format(msg))
 
+    @commands.command()
+    async def xkcd(self, ctx, xkcd_command: str = None):
+        """
+        Enjoy a nice xkcd comic with some strangers on the internet!
+        If no issue number is passed, returns a random xkcd.
+        Valid options are: nothing, 'latest' and a positive integer greater than one and at most equal to the latest issue number.
```
Comment (le-potate) [inline cogs/memes.py:181]: i'd write "or" instead of "and"

File: cogs/memes.py (hunk 6 above)
Comment (le-potate) [reply cogs/memes.py:199]: Actually why not send the latest instead if this happens

**Review by le-potate — CHANGES_REQUESTED** [review]

File: cogs/memes.py [hunk 9]
```diff
@@ -169,6 +173,58 @@ def pyramidy(n, m):
...
+        else:
+            try:
+                xkcd_num = int(xkcd_command)
+            except ValueError:
+                await ctx.send(
+                    f"invalid input: `{xkcd_command}` does not parse to an integer"
+                )
+                return
+            xkcd_req = requests.get(f"https://xkcd.com/{xkcd_num}")
+            if xkcd_num < 0:
```
Comment (le-potate) [inline cogs/memes.py:198]: < 1

**Review by le-potate — APPROVED** [review]
looks good to me

**Review by davidlougheed — APPROVED** [review]

**Pull request merged**